src/track/globaltrackcache.cpp

```diff
@@ -1,269 +1,379 @@
+}
+
+void GlobalTrackCacheResolver::initLookupResult(
         GlobalTrackCacheLookupResult lookupResult,
-        TrackRef trackRef,
-        TrackPointer pTrack)
-        : GlobalTrackCacheLocker(
-            std::move(moveable),
-            std::move(lookupResult),
-            std::move(trackRef),
-            std::move(pTrack)) {
+        TrackPointer&& strongPtr,
+        TrackRef&& trackRef) {
+    DEBUG_ASSERT(m_pInstance);
+    DEBUG_ASSERT(GlobalTrackCacheLookupResult::NONE == m_lookupResult);
+    DEBUG_ASSERT(!m_strongPtr);
+    m_lookupResult = lookupResult;
+    m_strongPtr = std::move(strongPtr);
+    m_trackRef = std::move(trackRef);
 }
 
-void GlobalTrackCacheResolver::updateTrackId(TrackId trackId) {
-    DEBUG_ASSERT(nullptr != m_pCacheMutex); // cache is still locked
+void GlobalTrackCacheResolver::initTrackIdAndUnlockCache(TrackId trackId) {
+    DEBUG_ASSERT(m_pInstance);
     DEBUG_ASSERT(GlobalTrackCacheLookupResult::NONE != m_lookupResult);
-    DEBUG_ASSERT(m_pTrack);
+    DEBUG_ASSERT(m_strongPtr);
     DEBUG_ASSERT(trackId.isValid());
-    m_trackRef = GlobalTrackCache::instance().updateTrackIdInternal(
-            m_pTrack,
-            m_trackRef,
-            trackId);
-    DEBUG_ASSERT(m_trackRef.getId() == trackId);
-    m_pTrack->initId(trackId);
-    DEBUG_ASSERT(createTrackRef(*m_pTrack) == m_trackRef);
+    if (m_trackRef.getId().isValid()) {
+        // Ignore initializing the same id twice
+        DEBUG_ASSERT(m_trackRef.getId() == trackId);
+    } else {
+        m_trackRef = m_pInstance->initTrackId(
+                m_strongPtr,
+                m_trackRef,
+                trackId);
+    }
+    unlockCache();
+    DEBUG_ASSERT(m_trackRef == createTrackRef(*m_strongPtr));
 }
 
-
-//static
-GlobalTrackCache* volatile GlobalTrackCache::s_pInstance = nullptr;
-
 //static
 void GlobalTrackCache::createInstance(GlobalTrackCacheEvictor* pEvictor) {
-    DEBUG_ASSERT(s_pInstance == nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(!s_pInstance);
     s_pInstance = new GlobalTrackCache(pEvictor);
 }
 
 //static
 void GlobalTrackCache::destroyInstance() {
-    DEBUG_ASSERT(s_pInstance != nullptr);
+    DEBUG_ASSERT(QApplication::instance()->thread() == QThread::currentThread());
+    DEBUG_ASSERT(s_pInstance);
     GlobalTrackCache* pInstance = s_pInstance;
+    // Reset the static/global pointer before entering the destructor
     s_pInstance = nullptr;
+    // Delete the singular instance
     delete pInstance;
 }
 
 //static
-void GlobalTrackCache::deleter(Track* pTrack) {
-    if (s_pInstance != nullptr) {
-        s_pInstance->evict(pTrack);
+void GlobalTrackCache::deleter(Track* plainPtr) {
+    DEBUG_ASSERT(plainPtr);
+    // Any access to plainPtr is forbidden!! Due to race condition
+    // this pointer might already have been deleted! This happens
```

Comment:
This sounds like a strong issue. The current implementation allows that a second reference counting object still lives in form of a number shared_pts, predicting holding a valid pointer. 
Currently this is not a real issue, because the last shared_ptr is only hold in the main thread guarded by the assertion in 
Library::afterEvictedTrackFromCache() but this will become one as soon as we change the code towards multi threading. 

The conceptual issue here is that the Cache mutex is tricked out because the reference counting atomic is decremented outside the guarded region. The time after decrement to 0 and locking the cache is the unsave time where second reference counting object can be created, even if we guard this by a second set of weak and shared pointers. 

This means also that the cache lock does not suite to protect reading from the file during writing. 
This works only, if we gurantee, that we have no second pointer out there in a second thread. 
So we need an atomic decrement inside the cache lock.

Reply:
I tried to explain this race condition during deletion many times now and why we need to handle it. The cache is not "tricked out", because this case is now properly handled by the eviction logic *after* the cache has been locked. If a new reference is created before the cache has been locked the deletion will be aborted. The cache also never creates a second track object instance for the same file (identified by its canonical location)!

Deleting track objects from different threads is and always was an issue, just without anyone noticing! The database updates are happening during the destruction of track objects and that didn't change with the introduction of the cache. I don't see how I introduced any new regression, but on the other hand I'm also not able to turn the whole architecture upside down.

Reply:
No, sorry this is a regression. In 2.1 it was done in the main thread guaranteed via a queued connection. 
https://github.com/mixxxdj/mixxx/blob/1.12/src/library/dao/trackdao.cpp#L1304

I think we should reintroduce that here, just to be save. 

The race that happens when saving is solved. I is proved by some nice unit tests. 
But the one, I am afraid of is an other.
If a destructor of the last reference is executed. You can still fetch the track from the cache. 
You can even fetch it n times and you will introduce a new reference counter each time. 
This means, it is not guaranteed that no one read the track during writing.
I am not sure what happens now since we have a temporary track, but original this was the issue that leads to sound artifacts and taglib crashes and was the reason to disable metadata write back.   

I originally thought this issue was gone with your latest changes, but it is back. Hopefully very unlikely now, but I prefer a rock solid solution. The second use counter version aims to be that one. 

I think if we move the temp file read into the lock and issue the DB update via a queued connection it should work. 

We may also consider to merge your or my one first and take some pressure from issue.